### ISSUE TITLE
I2C addresses not as expected.

### ISSUE THREAD

**Issue by Just-another-pleb** [atsage/node-red-contrib-LCD20x4-I2C#10]
Looking at your node it has addresses down in the 2x and 3x range.

Fair enough.

I did a bit of digging and found this link:
[https://www.raspberrypi.org/forums/viewtopic.php?t=64943](url) to look for I2C devices.

This is what I see:

![Screenshot from 2020-08-27 15-44-36](https://user-images.githubusercontent.com/39888280/91390258-ae88a180-e87c-11ea-8358-3b7e8dddb24f.png)

So it is going to be extremely difficult to get it working as I can't find an option to set the I2C address to 72.

It is a RasPi 3a and the display is:
[https://core-electronics.com.au/sparkfun-20x4-serlcd-black-on-rgb-3-3v.html](url)

Sorry, but I was (foolishly) hoping for an easy solution to this project.
The node looked like it was perfect to do what I am needing:   Send addressed messages to the display with x,y co-ordinates.
Ok, so it is line by line.   Maybe I can work around that part.

Bot to now, I can't get anything to be displayed.
I do see the "boot screen" of the display at power up.

**Comment by Barabba11** [comment]
What is boot screen? "SYSTEM START UP"?
If so your I2C connection works well, as the display and module, you simply don' tinput to module the right object.
If not that boot screen, then I suggest you to be sure about address of display, I don't remember but somewhere is in decimal form, somewhere else in hex. be sure

**Comment by Just-another-pleb** [comment]
It is fast.....    Something like "Spark fun" (something something.
(Just powered it up and it is probably not even 1 second displayed.)

It is hard to read as the back lite is not one and it is there for only a second.

Sorry.   Not helping I know.

What I showed you on the screen shot is hex.
Only talking/thinking.....

Can I edit the code for the node `lcddriver.py` and change the address to 72?
But I don't know how/if the drop down selector in `Node-Red` will like me doing that.

The display is *flashable* to a different address, but not by I2C - I believe.   I need to use serial.

**Comment by Barabba11** [comment]
you don't need to do that, you just need to edit the html file with the numbers of ports. Chances are the module uses decimal address, so change it there and select it from the menu. I think I did something like that and it worked, 2 years ago, I can't remember. 
If the Raspbian detect it it means it receives correctly answer from I2C device, in my opinion you need to focus on module. 
BUT! BUT! you can't exclude an electrical incompatibility because you are using sapberry bus which is 3,3volt on a 5Volt device, which may don't properly recognize high level "1", you better use a translator 3,3 to 5 Volt, as I did. Anyway you may try to help the bus to pull up a bit more to reinforce the level1, I sugges you to use 2 resistors of 2,2kohm to connect SCL and SDA to +3,3 volt (be sure you connect them to the raspberry 3,3 volt and not +5volt!!). If the line is not long, if it is longe you may use a lower value, like 1kohm. ANd this is better to put there resistors next to LCD, as a line termination. 
Please check your LCD, with a tester check scl and sda with its +5volt pin, and see if there is a termination resistor already on the LCD board. If it pull ups the bus at +5volt you may damage the raspberry, be careful.
I don't know if LCD may work at 3,3 volt too, in this case is safe, but it can be not stable. The best solution is a voltage translator, they are cheap

**Comment by Just-another-pleb** [comment]
No, this is a 3.3v device.   That was checked before purchase.   Alas I got *hit* with a couple of things but they are not critical.

Sorry to drag this up so long after you were done with it.
I am not getting the:
> you don't need to do that, you just need to edit the html file with the numbers of ports

Node-Red is `JavaScript` - yes, it constructs the page in HTML - I think.
But my idea is to get the node talking to the right address so anything else is .... *transparent* to me.
It is handled by the node.
So that would be either at the `Javascript` level or `python`.

**Comment by asage-me** [comment]
Edit LCD20x4-I2C.html, after line 36 add `<option value="0x72">0x72</option>`.  I'll try to update it later today.

**Comment by asage-me** [comment]
Actually, I thought that address sounded familiar.  You probably have a wiring issue.  That address is for a PWM controller https://i2cdevices.org/addresses/0x72

**Comment by Just-another-pleb** [comment]
Thanks.

Alas the machine in question is powered down just now, but I shall copy this and record it.
I'll try it tomorrow.
If that works, I will be so happy.

Alas then I will have to back engineer the node anyway, as I am wanting to send messages with an `x/y` position.   Not all 4 lines at once.
Yeah, I could *buffer* it and send it that way.
But that is *plan B*.

Thanks again though for the time and effort in replying to me.
It is appreciated.

**Comment by Just-another-pleb** [comment]
Edited the file as you said.
Set it to 0x72 and the display is now doing things.

Ok, using your example code where it sends 4 lines of text, I just get *junk* on the screen.
For the sake of eliminating things, I adjust the `scroll speed`.

I change it, close the window and `DEPLOY`.
The screen changes what it is showing.   But it is still a mess.   Nothing like what I am sending.

What does the *scroll speed* do?
Changing its value seems to change what is shown on the display.
So - theory -  if the speed is wrong the display gets the wrong message and therefore displays the wrong stuff.

**Comment by asage-me** [comment]
The scroll speed it just how fast the text scrolls across the display when it doesn't all fit.  What display are you using?  This node was made for a 20x4 HD44780 driven by a PCF8574.  Sounds like whatever you have at 0x72 just isn't compatible.

**Comment by Just-another-pleb** [comment]
[https://core-electronics.com.au/sparkfun-20x4-serlcd-black-on-rgb-3-3v.html](url)
I am sure it uses the same controller.

I hope it just needs a *tweak* to get all things working.

As said:  Now (with address 0x72 selected) when I inject your example, the display "changes".

But it won't change any more.

If I inject a new message payload, it also doesn't.  :(

But!  If I change the scroll speed, the display *blinks* and carrys on for a while.

So, to me that means the I2C is all hooked up and all that is good.

it is just the data being sent is ..... *wrong*.
And so should only need a tweak in the main code to get it working.

A bit more digging, and I am not exactly sure.   I may have to write some code and flash it to the display as it has its own "Arduino" chip on it.

It seems I could send data to it via the FTDI (serial) port, and get it displayed on screen as it.   not sure abut the I2C.
But I am also asking Sparkfun for help.

**Comment by Barabba11** [comment]
I?m sorry to say but your english isn't so good to let me understand your problems, what exactly means "blinks and carrys on for a while"? A video would be appreciated.
I suggest you to introduce pullup resistors on I2C line, you can't imagine how much they are important

**Comment by Just-another-pleb** [comment]
Sorry.   When I change the `scroll time` and `DEPLOY` the flow, the display shows random stuff for a while.

To me it looks like it is receiving data but not *valid* data.   More just random stuff.

The I2C cable is only .... 5CM long.

I think that there is a problem with what the display does at this point.

I was hoping that *out of the box* I could talk to it on any of the ports.  I2C, SPI or FTDI.
What I have seen on this link:
[https://www.youtube.com/watch?time_continue=192&v=IKiTNX9SVqw&feature=emb_logo](url)
I have to program the processor if I want to use either the I2C or SPI bus.

Only the FTDI bus work where I can send stuff to be displayed now.
To be able to use the I2C or SPI I have to program the onboard AVR chip.

I'm guessing your node is only expecting a bare display.  Not one with an onboard AVR.

I am still not 100% sure.

Could you (please) watch it and tell me how you understand the video to mean in what the display I have does?

**Comment by Barabba11** [comment]
it's hard to guess what is wrong there, acording to guide you don't have to setup nothing hardware to tell you are using I2C bus, and if you see this 72 (which disappear when you remove LCD looks they are communicationg. Probably there is a filter between LCD processor and the i2C bus. This module is intended to talk directly with LCD processor, but here there is an ATmega processon that enables all serial interfaces.. 
try to pull up the lines, if it may help, if not probably they aren't compatible. write them on github and ask
Maybe you would better buy a cheap i2c lcd with 5 volt levels and put a voltage translator, all it will be very cheap

**Comment by Just-another-pleb** [comment]
Hey, thanks for the time and effort.

I have got the display working - but gee it was convoluted in doing it.

I had to get some *qwiic* package from Sparkfun and install it.

It works mostly.

But when I try to adjust the backlight colour it goes crazy.

It prints messages on the screen of the backlight colour when I didn't ask, and the colour it says is not the colour of the screen.

So I think I'd better leave you alone as it is way beyond your scope.
